lib/logstash/filters/kubernetes_metadata.rb

```diff
@@ -0,0 +1,290 @@
+  #   - metadata['log_format_stdout']
+  # This is completely up to the user to supply and use in their logstash config and will not
+  # be used anywhere else
+  config :default_log_format, :validate => :string, :default => "default"
+
+  public
+  def register
+    @logger.debug("Registering Kubernetes Filter plugin")
+    self.lookup_cache ||= LruRedux::ThreadSafeCache.new(1000, 900)
+    @logger.debug("Created cache...")
+  end
+
+  # this is optimized for the single container case. it caches based on filename to avoid the
+  # filename munging on every event.
+
+  public
+  def filter(event)
+    path = event.get(@source)
+
+    # Ensure that the path parameter has been defined so that we can find the required metadata
+    if (path.nil? || path.empty?)
+      event.tag("_kubeparsefailure")
+      return
+    end
+
+    @logger.debug("Log entry has source field, beginning processing for Kubernetes")
+
+    metadata = {}
+    cached_metadata = lookup_cache[path]
+    file_metadata = get_file_info(path)
+
+    # If we were unable to extract metadata from the file name, return
+    return unless file_metadata
+
+    if cached_metadata
+      metadata = file_metadata.merge(cached_metadata)
+    else
+      @logger.debug("Trying to get kubernetes file info, it was not cached");
+      @logger.debug("kubernetes file info got: #{metadata}")
+
+      pod = file_metadata['pod']
+      namespace = file_metadata['namespace']
+      name = file_metadata['container_name']
+
+      return unless pod and namespace and name
+
+      if data = get_kubernetes(namespace, pod)
+        metadata = file_metadata.merge(data)
+        set_log_formats(metadata)
+        lookup_cache[path] = metadata
+      end
+      @logger.debug("metadata within lookup_cache[path]: #{metadata}")
+    end
+
+
+    @logger.debug("metadata after unless lookup_cache[path] is: #{metadata}")
+    @logger.debug("config after unless lookup_cache[path] is: #{config}")
+
+    event.set(@target, metadata)
+    return filter_matched(event)
+  end
+
+  def set_log_formats(metadata)
+    begin
+      #return if metadata['annotations'].empty?
+
+      format = {
+        'stderr' => @default_log_format,
+        'stdout' => @default_log_format
+      }
+      a = metadata['annotations']
+      n = metadata['container_name']
+
+      # check for log-format-<stream>-<name>, log-format-<name>, log-format-<stream>, log-format
+      # in annotations
+      %w{ stderr stdout }.each do |t|
+        [ "log-format-#{t}-#{n}", "log-format-#{n}", "log-format-#{t}", "log-format" ].each do |k|
+          if v = a[k]
+            format[t] = v
+            break
+          end
+        end
+      end
+
+      metadata['log_format_stderr'] = format['stderr']
+      metadata['log_format_stdout'] = format['stdout']
+      @logger.debug("kubernetes metadata", metadata: metadata)
+
+    rescue => e
+      @logger.warn("Error setting log format. Please check log-format annotation.", error => e.to_s)
```

Comment:
Probably would be good to know. I went ahead and added a `_kubelogformatparsefailure` tag before this warn.